core/src/main/java/com/linecorp/armeria/server/auth/AuthServiceBuilder.java

```diff
@@ -140,6 +141,7 @@ public AuthServiceBuilder addOAuth2(Authorizer<? super OAuth2Token> authorizer,
             return authorizer.authorize(ctx, token);
         };
         add(requestAuthorizer);
+        setAuthorizerHandlers(authorizer);
```

Comment:
What would happen if a user adds more than one authorizer? Wouldn't each authorizer need to have its own success/failure handler pair?

Reply:
I don't think it makes sense or even possible to chain success/failure handlers. Unlike authorizers, which represent decision making points, handlers represent concrete actions, like `AuthSuccessHandler successHandler = Service::serve;`. How can this action get chained or replicated?
The problem is that for some protocols like OAuth 2 the action (concretely failureHandler) is bound to the decision making. That's why it made sense to add handlers to an Authorizer as you've suggested. However, Authorizer chaining brings ambiguity to that.

IMHO Authorizer chaining is quite ambiguous and dangerous even without that, at least in the way the logic implemented right now - it goes against some basic authorization principles. Solution implemented today represents a permissive mechanism where two conflicting policies (represented by Authorizers) may permit the client even if one of the policies explicitly says NO. It should work the other way around - it should permit the client only if ALL policies said YES.

Do you know of the use cases when it's required to have authorizers chaining? Could it be solved via multiple AuthServices instead?

Reply:
Hey, thanks for the feedback and sorry for giving a poor suggestion.

The authorizer chaining was introduced to handle the case where a service has to support different protocols, e.g. OAuth1, OAuth2 and authorize by source IP address, etc.

I think it can't be solved by having multiple `AuthService`s in the pipeline because the first `AuthService` will have to authorize the request to make the second `AuthService` evaluated.

Ideally, we could change the return value of `Authorizer.authorize()` so it can represent one of the following states:

- AUTHORIZED - `successHandler` will handle the request.
- UNAUTHORIZED - `failureHandler` will handle the request.
- NEXT - ask the next authorizer. If reached at the last authorizer, `failureHandler` will handle the request.

However, we can't do this due to backward compatibility.

How about doing this until we release Armeria 2 (maybe late 2021?) as a compromise?

- When authorizer returns `true`, use the authorizer's `successHandler`.
- When authorizer returns `false`, try the next authorizer. (i.e. retain the current `AuthService` behavior.)
- When the last authorizer returns `false`, use the last authorizer's `failureHandler`.

Reply:
Sure, let me think about it. However, I see some problem immediately - this will break the tie between the Authorizer and Success/FailureHandlers. Following your proposal, in case when there are two Authorizers, then AuthService could engage SuccessHandler associated with the first Authorizer and engage FailureHandler associated with the second Authorizer. It could be messy. If we keep Authorizer chaining, we have to disassociate handlers from the Authorizers (i.e. revert the last changes I did).

Reply:
Yeah, I thought about that problem where only the last failure handler will be used. I see a few ways to address this:

- Document the behavior (only the last authorizer's `failureHandler` will be used).
- Make sure all authorizers except the last one did not override `failureHandler()` method. Raise an exception if not.
- Revert the change that added `successHandler()` and `failureHandler()`.
  - But we need to how we can make `OAuth2TokenIntrospectionAuthorizer` to configure `AuthServiceBuilder`. It'd be nice if `AuthService.onFailure()` is automatically called when `OAuth2TokenIntrospectionAuthorizer` for example.

Which would be the best?

Reply:
I think if we want to keep Authorizer chaining in place, we should make following adjustments:

As you pointed, ideally, we should change Authorizer.authorize() signature to return {AUTHORIZED, UNAUTHORIZED, NEXT}. However, I’m sure when would the authorizer return NEXT really – as you know most on the authorization systems always return NO_ACCESS, no matter whether the user specified wrong credentials, has insufficient rights or simply absent in the authorization DB altogether.

Therefore, we should change chaining semantics from “going next on failure” to “going next on success”. I.e. reverse the current AuthService behavior:

- When authorizer returns false, use the authorizer's failure handler.
- When authorizer returns true, try the next authorizer.
- Reaching the end of the success chain engages common success handler.

This will require minimal changes to the API, however, the chaining will work differently.

On the implementation side the AuthService will still have common success and default failure handlers. Each Authorizer may be permitted to define only custom failure handler (which will serve as authorization sequence terminator). If an Authorizer provides its own handler, the AuthService will use that to terminate, but if the Authorizer does not provide a custom handler, the AuthService will use the default one.

What do you think?

Reply:
We could make it even more complex and introduce a “chaining mode” for the chain to work: {`AND`, `OR`}. The Authorizer would remain binary (`true/false`), but the AuthService to engage next Authorizer and terminate the sequence either on failure (`AND`) or on success (`OR`). The `AND` case I have described above. The `OR` case would look like the following:

The AuthService invokes the first Authorizer form the chain. If it succeeds, it terminates the chain and invokes common success handler. If it fails, the AuthService tries the next Authorizer, and so on. At the end of the chain, if none of the Authroizers returned success, the AuthService selects the Authorizer to handle the failure handler (typically first or last, or even designated one). The AuthService always engages matching failure handler from the selected Authorizer. If an authorizer does not define a failure handler, the AuthService selects the one that does or engages default failure handler.
I think `AND` mode should be the default one, as it matches security practices better. `OR` mode should be specified explicitly.

To summarize the interface changes:

- AuthService will have `successHandler` as before and existing `failureHandler` will rename to `defaultFailureHandler`.
- AuthService will have a `chainMode` that defaults to `AND`.
- AuthService will have an additional configuration on which Authorizer to select in `OR` mode with following policies: {`FIRST`, `LAST`, `FIRST_WITH_HANDLER`, `LAST_WITH_HANDLER`, `SELECTED`}.
- Authorizer will have `failureHandler` with default implementation to NULL, and will no longer have `successHandler`.

Reply:
Thanks for the idea. I have a few comments.

> `AuthService` will have a `chainMode` that defaults to `AND`.

I guess we can't default to `AND` in this case because it will be a breaking change, because `AuthService` and `Authrizer` are unfortunately (?) not unstable. For now, could we just stay with the current behavior of 'success if true, next if false'? I think it's fine as long as a user has one authorizer. We can redesign this properly in Armeria 2 or defer the changes like introducing `chainMode`, unless this problem blocks you from implementing OAuth2. This way, we could merge your work sooner without making too much changes in the core API.

> Authorizer will have `failureHandler` with default implementation to NULL, and will no longer have `successHandler`.

Wouldn't it be better having `successHandler` for each authorizer? An authorizer might want to perform some action specific to its corresponding auth mechanism, e.g. updating metrics, redirecting to an intermediary page, etc.

Reply:
Ok, I will implement it in the following way "for now" following `OR` mode.
The `AuthService` invokes the first `Authorizer` form the chain. If it succeeds, it terminates the chain and invokes associated `successHandler`. If `successHandler` not defined for given `Authorizer`, it will invoke default/common `successHandler`. If the `Authorizer` fails (returns `false`), the `AuthService` tries the next `Authorizer`, and so on. At the end of the chain, if none of the `Authroizers` returned success, the `AuthService` selects the `Authorizer` to handle the `failureHandler` (`LAST_WITH_HANDLER` - the last one that has specific `failureHandler` defined). The `AuthService` always engages matching `failureHandler` from the selected `Authorizer`. If none of the `Authorizer`s define a `failureHandler`, the AuthService selects engages default/common `failureHandler`.

Summary of the interface changes:
- `AuthService` will have `successHandler` and `failureHandler` as before - these will play a role of default/common handlers.
- `AuthService` will have `LAST_WITH_HANDLER` `Authorizer` selection strategy hardcoded.
- `Authorizer` will have `failureHandler` and `successHandler` set to `NULL` by default, as before.

This solution will be fully backward compatible and the new changes will mostly take place in `AuthService#serve()` method.

Reply:
Excellent. Thank you so much for your understanding. I agree that this is not ideal but it will help us move forward. Looking forward to this PR being merged and your follow-up PRs! :bow:

Reply:
Pushed the requested changes as described above. Had to extract chaining logic into dedicated `AuthorizerChain` and extract `HttpRequest` token authorization into dedicated `DelegatingHttpRequestAuthorizer` classes. Made a minor correction to existing test case (`AuthorizerTest#orElseToString()`).
Sorry for the delays - got lot's of work lately.